Use Chrome for Testing Canary for CI runs

DanielRyanSmith opened the pull request:
This change updates CI tests to run Chrome for Testing Canary instead of Chrome Dev.

DanielRyanSmith committed:
tc runs chrome canary

DanielRyanSmith committed:
comments and cleanup

DanielRyanSmith committed:
Move directory for integration tests

jcscottiii commented on tools/ci/run_tc.py:
```diff
@@ -146,25 +146,20 @@ def start_dbus():
-    run(["sudo", "gdebi", "-qn", "/tmp/%s" % deb_archive])
-
+    # Interpret experimental channels as Canary.
+    if channel in ("experimental", "nightly", "canary"):
```
What happens for the stable and beta cases now?

DanielRyanSmith replied:
This is just a confirmation that we're interpreting "experimental" and "nightly" as "canary", since that is where the files will be when installing. Beta and stable should still be handled normally. 🙂

jgraham replied:
I'm confused in the same way @jcscottiii is. It looks like this removes all support for installing non-Canary versions of Chrome. But are those not still required for stable/beta runs in CI?

Also, for installation that doesn't involve a debian package we shouldn't have any code in `run_tc.py`, but should instead pass `--install-browser` to `wpt run` (look at what we're doing for Firefox). Ideally this environment setup wrapper wouldn't need to know anything about which browser we're going to end up running.

DanielRyanSmith replied:
Ah, I'm sorry to both of you 😓 - I missed that the `channel` variable is not being used in the `./wpt install` statement, and canary is hardcoded in. I thought the confusion was with the channel interpretation logic. This is now fixed.

As for Firefox, I believe it is not installed using `wpt run` for the integration tests, which also had me scratching my head a bit looking to mimic the implementation - but it turns out it the installation is [specified in the Dockerfile via apt-get](https://github.com/web-platform-tests/wpt/blob/master/tools/docker/Dockerfile#L51-L56)

jgraham replied:
> As for Firefox, I believe it is not installed using wpt run for the integration tests

Huh, if so that seems very much like a bug / regression. The only reason we install Firefox in the Docker image is to ensure we have all the right dependencies. Maybe we should remove the Firefox binary to stop that happening.

foolip commented:
Will we have Dev and Canary runs in parallel? We haven't decided yet what the end date of the Interop 2023 dashboard will be, so we might need Chrome Dev runs until it's frozen.

If the switch is made and Chrome Dev no longer runs, I suppose we'd just start using the Canary runs assuming they're labeled experimental, but the dashboard uses the Chrome Dev name and logo.

If both run in parallel the current scoring code will probably pick whichever finished running last, which is likely to be different on different days.

DanielRyanSmith committed:
run Chrome dev alongside Canary

DanielRyanSmith committed:
add Chrome Dev download

DanielRyanSmith committed:
remove extra chrome download

DanielRyanSmith commented:
> Will we have Dev and Canary runs in parallel? We haven't decided yet what the end date of the Interop 2023 dashboard will be, so we might need Chrome Dev runs until it's frozen.

Yes, we had talked about this, and I had not added both to run in parallel in this PR. Sorry about that! I've made changes so that Dev should also run alongside Canary until we decide to remove it. 🙂

jgraham commented on tools/wpt/browser.py:
```diff
@@ -1230,19 +1230,12 @@ def find_binary(self, venv_path=None, channel=None):
-                name += "-unstable"
-            # No Canary on Linux.
-            return which(name)
+            return which("chrome")
```
Does this also work with other channels?

DanielRyanSmith replied:
Yes - all versions of the binary are named "chrome" using Chrome for Testing on Linux.

jgraham commented on tools/ci/run_tc.py:
```diff
@@ -146,25 +146,20 @@ def start_dbus():
+        path = os.path.join(os.getcwd(),
+                            f"_venv3/browsers/{channel}/chrome-linux64",
+                            file_name)
+        run(["sudo", "mv", path, "/usr/bin"])
```
Is this actually required or is it just a simplification so that `which` will pick things up later? In the latter case I'd strongly prefer not to do this.

DanielRyanSmith replied:
As far as I could tell with the integration tests (which this setup is for), The venv folder is removed or rebuilt at some point after this step, and the browser binary cannot be found. `/usr/bin` is where the previous Chrome Dev binary was installed to, and where the Firefox binary exists as well.

jgraham commented on tools/ci/run_tc.py:
```diff
@@ -146,25 +146,20 @@ def start_dbus():
+        channel = "canary"
+    run(["sh", "-c", "./wpt install --channel canary chrome browser"])
+
+    # Chrome for Testing needs additional files in the download directory to run all features.
```
I don't really know what this comment is trying to say. The code doesn't seem to end up adding additional files, but moving the downloaded ones?

DanielRyanSmith replied:
When I was setting this up and running the integration tests, I was moving the `chrome` binary only to the new directory, which was causing unintended side effects and a failure for the browser tests to run. The problem was that the binary relies on (at least some of) the other files downloaded to also exist in the same directory for binary to detect them and run properly.

The comment could definitely be worded better - I'll update it. 😅 Also, if there is a better or more preferable approach to doing this, I'd be glad to update this.

jgraham reviewed (changes requested):
Is there actually a difference in Interop scores between dev and canary that isn't explained simply by the fact that Canary is slightly newer code? If there is I think we should hold this until Interop 2024 starts, if there isn't then I think we should just transition directly to Canary. Although running them side-by-side for a while feels like good practice, I think it's adding significant extra load and extra complexity for very little gain.

DanielRyanSmith committed:
Update run_tc.py

DanielRyanSmith commented:
> Is there actually a difference in Interop scores between dev and canary that isn't explained simply by the fact that Canary is slightly newer code? If there is I think we should hold this until Interop 2024 starts, if there isn't then I think we should just transition directly to Canary. Although running them side-by-side for a while feels like good practice, I think it's adding significant extra load and extra complexity for very little gain.

I will look into the major differences after obtaining some run data now that #43805 is merged.

After addressing the comments here, I think it is probably best to break this PR up a bit more, since it is trying to make multiple, and somewhat unrelated changes. I'll split this up into PRs regarding:
1. Obtaining necessary Chrome Canary runs on wpt.fyi.
2. Changing the non-integration test CI runs to use Chrome for Testing.
3. Changing the integration tests to use Chrome for Testing.

DanielRyanSmith commented:
Update: I triggered Chrome Dev and Canary to run with the same WPT changes, and [here are the diffed results](https://wpt.fyi/results/?diff&filter=ADC&run_id=5203276210896896&run_id=5114590957928448)
(Note that Chrome Canary is using the Chromium logo, which is fixed in a newly submitted PR).

foolip commented:
@DanielRyanSmith based on that, here are some possible regressions:

https://wpt.fyi/results/?diff&filter=ADC&q=seq%28status%3Apass%20status%3A%21pass%29&run_id=5203276210896896&run_id=5114590957928448
https://wpt.fyi/results/?diff&filter=ADC&q=seq%28status%3Aok%20status%3A%21ok%29&run_id=5203276210896896&run_id=5114590957928448

But flakiness is likely to account for a large amount of the differences. I'd suggest triggering two runs for each configuration, and then using a query like `seq(status:pass status:pass status:!pass status:!pass)` to filter out most of it.

jgraham commented:
Just by eye I think you can see that there are a few features which aren't enabled in the new configuration and a large group of tests are now failing, but that most of the cases where it's a few tests in a larger feature there's a history of the test being intermittent (I also don't think that query will show all relevant cases, since it would miss anything that's gone from e.g. PASS to FAIL).

DanielRyanSmith committed:
Merge branch 'master' into 2023-10-05_cft-canary
